zebra-chain/src/value_balance.rs

```diff
@@ -0,0 +1,297 @@
+            sapling: (self.sapling - rhs.sapling)?,
+            orchard: (self.orchard - rhs.orchard)?,
+        })
+    }
+}
+impl<C> std::ops::Sub<ValueBalance<C>> for Result<ValueBalance<C>, ValueBalanceError>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+    fn sub(self, rhs: ValueBalance<C>) -> Self::Output {
+        self? - rhs
+    }
+}
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest::prelude::*;
+#[cfg(any(test, feature = "proptest-impl"))]
+impl Arbitrary for ValueBalance<NegativeAllowed> {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (
+            any::<Amount>(),
+            any::<Amount>(),
+            any::<Amount>(),
+            any::<Amount>(),
+        )
+            .prop_map(|(transparent, sprout, sapling, orchard)| Self {
+                transparent,
+                sprout,
+                sapling,
+                orchard,
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+#[cfg(any(test, feature = "proptest-impl"))]
+impl Arbitrary for ValueBalance<NonNegative> {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+        )
+            .prop_map(|(transparent, sprout, sapling, orchard)| Self {
+                transparent,
+                sprout,
+                sapling,
+                orchard,
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    proptest! {
+        #[test]
+        fn test_add(
+            value_balance1 in any::<ValueBalance>(),
+            value_balance2 in any::<ValueBalance>())
+        {
+            zebra_test::init();
+
+            let transparent = value_balance1.transparent + value_balance2.transparent;
+            let sprout = value_balance1.sprout + value_balance2.sprout;
+            let sapling = value_balance1.sapling + value_balance2.sapling;
+            let orchard = value_balance1.orchard + value_balance2.orchard;
+
+            match (transparent, sprout, sapling, orchard) {
+                (Ok(transparent), Ok(sprout), Ok(sapling), Ok(orchard)) => prop_assert_eq!(
+                    value_balance1 + value_balance2,
+                    Ok(ValueBalance {
+                        transparent,
+                        sprout,
+                        sapling,
+                        orchard,
+                    })
+                ),
+                _ => {
+                    let error_string = format!("{:?}", (value_balance1 + value_balance2));
+                    prop_assert_eq!(error_string.contains("AmountError"), true)
+                }
```

Comment:
You might find the `matches!` macro helpful here:
https://doc.rust-lang.org/std/macro.matches.html

```suggestion
                _ => prop_assert!(matches!(result, Err(ValueBalanceError::AmountError(_)))),
```

Reply:
Awesome, i was looking for this solution since yesterday :)